File: elixir/all-your-base/lib/all_your_base.ex
defmodule AllYourBase do
  defmodule DigitsOps do
    @moduledoc false

    @smallest_digit 0

    # Returns a closure that checks whether the given digit is
    # valid for the base in which it was constructed.
    @spec base_checker(integer) :: (integer -> boolean)
    def base_checker(base) do
      fn digit ->
        digit >= @smallest_digit and digit < base
      end
    end

    # Returns a closure that returns the value of a given digit at the
    # given position based on the base used to construct the closure.
    @spec value_getter(integer) :: ({integer, integer} -> integer)
    def value_getter(base) do
      fn {digit, position} ->
        digit * Integer.pow(base, position)
      end
    end
  end

  defmodule PositionalUtils do
    @moduledoc false

    @min_digits 1
    @min_defined_log 1

    # Calculates and returns the number of digits required to represent
    # the given value in the given base.
    @spec digits_count(integer, integer) :: integer
    def digits_count(value, base)

    # special logs value
    def digits_count(value, _) when value <= @min_defined_log,
      do: @min_digits

    # common case
    def digits_count(value, base) do
      log_in_base = :math.log(value) / :math.log(base)

      log_in_base
      |> Float.ceil()
      |> trunc()
    end

    # Checks and returns whether the digits are valid with a number in the given base.
    @spec valid_number?([integer], integer) :: boolean
    def valid_number?(digits, base),
      do: Enum.all?(digits, AllYourBase.DigitsOps.base_checker(base))
  end

  @smallest_base 2

  @doc """
  Given a number in input base, represented as a sequence of digits, converts it to output base,
  or returns an error tuple if either of the bases are less than 2
  """
  @spec convert(list, integer, integer) :: {:ok, list} | {:error, String.t()}
  def convert(digits, input_base, output_base)

  # invalid input base
  def convert(_, input_base, _) when input_base < @smallest_base,
    do: {:error, "input base must be >= 2"}

  # invalid output base
  def convert(_, _, output_base) when output_base < @smallest_base,
    do: {:error, "output base must be >= 2"}

  # without digits is treated as zero in all bases
  def convert([], _, _),
    do: {:ok, [0]}

  # attempt to convert to the same base
  def convert(digits, same_base, same_base) do
    if PositionalUtils.valid_number?(digits, same_base) do
      {:ok, digits}
    else
      {:error, "all digits must be >= 0 and < input base"}
    end
  end

  # commom case
  def convert(digits, input_base, output_base) do
    if !PositionalUtils.valid_number?(digits, input_base) do
      {:error, "all digits must be >= 0 and < input base"}
    else
      {
        :ok,
        digits
        |> digits_to_value(input_base)
        |> value_to_digits(output_base)
      }
    end
  end

  # Obtains the real value of the positional representation
  # given by the digits and the base.
  @spec digits_to_value([integer], integer) :: integer
  defp digits_to_value(digits, base) do
    digits
    |> Enum.reverse()
    |> Stream.with_index()
    |> Stream.map(DigitsOps.value_getter(base))
    |> Enum.reduce(&+/2)
  end

  # Converts the given value to its positional representation
  # in the given base.
  @spec value_to_digits(integer, integer) :: [integer]
  defp value_to_digits(value, base) do
    (PositionalUtils.digits_count(value, base) - 1)..0//-1
    |> Stream.map(fn pow -> Integer.pow(base, pow) end)
    |> Stream.map(fn base_pows -> Integer.floor_div(value, base_pows) end)
    |> Enum.map(fn val -> rem(val, base) end)
  end
end

File: cpp/all-your-base/all_your_base.cpp
#include "all_your_base.h"

#include <algorithm>
#include <cmath>
#include <stdexcept>

namespace all_your_base {

/**
 * @typedef value_t
 * @brief Represents the decimal value of a number, used for intermediate conversions.
 */
using value_t = uint64_t;

/**
 * @brief Converts a number from a given base to its decimal value.
 *
 * @param base The base of the number.
 * @param number The number represented as a vector of digits.
 * @return The decimal value of the number.
 */
value_t get_value(base_t base, const number_t &number);
/**
 * @brief Converts a decimal value to a number in a specified base.
 *
 * @param value The decimal value to convert.
 * @param base The target base.
 * @return A vector of digits representing the number in the target base.
 */
number_t value_to_base(value_t value, base_t base);
/**
 * @brief Calculates the decimal value of a single digit at a specific position.
 *
 * @param digit The digit value.
 * @param base The base of the number.
 * @param position The position of the digit (right to left, starting from 0).
 * @return The decimal value of the digit at the given position.
 */
value_t value_of_digit(digit_t digit, base_t base, size_t position);
/**
 * @brief Validates if a number's digits are correct for its base.
 *
 * Checks if every digit in the number is less than its base.
 *
 * @param base The base of the number.
 * @param number The number as a vector of digits.
 * @return `true` if the number is valid, `false` otherwise.
 */
bool is_valid_number(base_t base, const number_t &number);

/**
 * @var MIN_BASE The minimum valid base for conversions.
 */
const base_t MIN_BASE = 2;

/**
 * @brief Converts a number from one base to another.
 *
 * This function first converts the input number to a decimal value, then converts
 * that decimal value to the output base. It includes extensive validation for
 * the bases and the input digits.
 *
 * @param input_base The base of the input number. Must be >= 2.
 * @param input A vector of digits representing the number in `input_base`.
 * @param output_base The desired base for the output number. Must be >= 2.
 * @return A vector of digits representing the converted number in `output_base`.
 * @throws std::invalid_argument If an input base or output base is less than 2,
 * or if any digit in the input number is not valid for the input base.
 */
number_t convert(base_t input_base, const number_t &input, base_t output_base) {
  if (input_base < MIN_BASE || output_base < MIN_BASE) {
    throw std::invalid_argument("");
  }
  if (!is_valid_number(input_base, input)) {
    throw std::invalid_argument("");
  }
  if (input_base == output_base) {
    return input;
  }
  const value_t value = get_value(input_base, input);
  return value_to_base(value, output_base);
}

/**
 * @brief Converts a number from a given base to its decimal value.
 *
 * It processes the digits of the number from right to left, multiplying each
 * digit by the corresponding power of the base and summing the results.
 *
 * @param base The base of the number.
 * @param number The number represented as a vector of digits.
 * @return The decimal value of the number.
 */
value_t get_value(base_t base, const number_t &number) {
  value_t value = 0;

  for (size_t digit_count = 0; digit_count < number.size(); digit_count++) {
    const size_t index = number.size() - digit_count - 1;
    value += value_of_digit(number[index], base, digit_count);
  }

  return value;
}

/**
 * @brief Calculates the decimal value of a single digit at a specific position.
 *
 * @param digit The digit value.
 * @param base The base of the number.
 * @param position The position of the digit (right to left, starting from 0).
 * @return The decimal value of the digit at the given position.
 */
value_t value_of_digit(digit_t digit, base_t base, size_t position) {
  return digit * std::pow(base, position);
}

/**
 * @brief Converts a decimal value to a number in a specified base.
 *
 * It uses the repeated division method, storing the remainders as digits.
 * The resulting digits are reversed to form the correct number.
 *
 * @param value The decimal value to convert.
 * @param base The target base.
 * @return A vector of digits representing the number in the target base.
 */
number_t value_to_base(value_t value, base_t base) {
  number_t result;

  for (value_t current = value; current > 0; current /= base) {
    result.push_back(current % base);
  }

  std::reverse(result.begin(), result.end());

  return result;
}

/**
 * @brief Validates if a number's digits are correct for its base.
 *
 * Checks if every digit in the number is less than its base.
 *
 * @param base The base of the number.
 * @param number The number as a vector of digits.
 * @return `true` if the number is valid, `false` otherwise.
 */
bool is_valid_number(base_t base, const number_t &number) {
  return std::all_of(number.begin(), number.end(),
                     [base](digit_t digit) { return digit < base; });
}

}  // namespace all_your_base